File: 2024/day21.c
// Precomputed panel moves using day21.rb

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <stdint.h>

#include "lib.c"


typedef struct {
    char direction;
    int magnitude;
} movement;

typedef struct {
    int count;
    movement movements[3];
} movements;

typedef struct {
    int count;
    movements options[2];
} movement_options;

static movement_options NUMBER_PANEL_MOVES[256][256] = { 0 };
static movement_options DIRECTION_PANEL_MOVES[256][256] = { 0 };


void initialise_movement_tables() {
    #include "day21_table.h"
}


unsigned int cache_key(movements desired_move, int depth) {
    unsigned int result = 0;

    // 5 bits per movement for a total possible of 15
    for (int i = 0; i < desired_move.count; i++) {
        result <<= 3;

        switch (desired_move.movements[i].direction) {
        case '^':
            result |= 0;
            break;
        case '<':
            result |= 1;
            break;
        case 'v':
            result |= 2;
            break;
        case '>':
            result |= 3;
            break;
        case 'A':
            result |= 4;
            break;
        }

        result <<= 2;
        result |= (unsigned int)desired_move.movements[i].magnitude;
    }

    // A further 5 bits for the depth (up to 25 in real life)
    result <<= 5;
    result |= (unsigned int) depth;

    return result;
}

static int64_t *CACHE;

int64_t lowest_directional_cost(movements desired_move, int depth) {
    unsigned int key = cache_key(desired_move, depth);

    if (CACHE[key] > 0) {
        return CACHE[key];
    }

    int64_t result = 0;

    if (depth == 0) {
        for (int i = 0; i < desired_move.count; i++) {
            result += desired_move.movements[i].magnitude;
        }
    } else {
        char current_cell = 'A';

        for (int i = 0; i < desired_move.count; i++) {
            int64_t best_cost = INT64_MAX;

            movement m = desired_move.movements[i];

            movement_options options = DIRECTION_PANEL_MOVES[(int)current_cell][(int)m.direction];

            for (int option_idx = 0; option_idx < options.count; option_idx++) {
                int64_t cost = lowest_directional_cost(options.options[option_idx], depth - 1);
                cost += (m.magnitude - 1);

                if (cost < best_cost) {
                    best_cost = cost;
                }
            }

            result += best_cost;
            current_cell = m.direction;
        }
    }

    CACHE[key] = result;

    return result;
}


int64_t lowest_cost(char *code, int directional_panel_count) {
    char current_cell = 'A';

    int64_t total_cost = 0;

    for (char *digit = code; *digit; digit++) {
        int64_t best_cost = INT64_MAX;

        movement_options options = NUMBER_PANEL_MOVES[(int)current_cell][(int)*digit];

        assert(options.count > 0);

        for (int option_idx = 0; option_idx < options.count; option_idx++) {
            int64_t cost = lowest_directional_cost(options.options[option_idx], directional_panel_count);

            if (cost < best_cost) {
                best_cost = cost;
            }
        }

        current_cell = *digit;
        total_cost += best_cost;
    }

    return total_cost;
}

void part1() {
    #include "input_files/day21.txt"

    int64_t total_complexity = 0;

    for (int code_idx = 0; code_idx < alen(codes_to_enter); code_idx++) {
        int64_t cost = lowest_cost(codes_to_enter[code_idx], 2);
        int64_t numeric = 0;

        for (char *c = codes_to_enter[code_idx]; *c; c++) {
            if (*c >= '0' && *c <= '9') {
                numeric *= 10;
                numeric += *c - '0';
            }
        }

        total_complexity += (cost * numeric);
    }

    printf("Part 1: Total complexity: %ld\n", total_complexity);
}

void part2() {
    #include "input_files/day21.txt"

    int64_t total_complexity = 0;

    for (int code_idx = 0; code_idx < alen(codes_to_enter); code_idx++) {
        int64_t cost = lowest_cost(codes_to_enter[code_idx], 25);
        int64_t numeric = 0;

        for (char *c = codes_to_enter[code_idx]; *c; c++) {
            if (*c >= '0' && *c <= '9') {
                numeric *= 10;
                numeric += *c - '0';
            }
        }

        total_complexity += (cost * numeric);
    }

    printf("Part 2: Total complexity: %ld\n", total_complexity);
}



int main() {
    initialise_movement_tables();

    // Enough space based on the bits used by cache_key (20 bits).
    //
    // Too lazy to use a proper hash table.  Let's just call it perfect hashing and
    // pretend we're virtuous.
    size_t cache_size = 1 << 20;

    CACHE = malloc(cache_size * sizeof(int64_t));

    memset(CACHE, 0, cache_size * sizeof(int64_t));
    part1();

    memset(CACHE, 0, cache_size * sizeof(int64_t));
    part2();
}

File: 2024/day21.rb

NUMBER_PANEL = [["7", "8", "9"],
                ["4", "5", "6"],
                ["1", "2", "3"],
                ["X", "0", "A"]]

DIRECTION_PANEL = [["X", "^", "A"],
                   ["<", "v", ">"]]


def precompute_best_moves(panel)
  result = {}

  width = panel[0].length
  height = panel.length

  (0...width).each do |from_col|
    (0...height).each do |from_row|
      from_cell = panel.fetch(from_row).fetch(from_col)

      (0...width).each do |to_col|
        (0...height).each do |to_row|
          to_cell = panel.fetch(to_row).fetch(to_col)

          moves = []

          if from_cell == 'X' || to_cell == 'X'
            moves << []
            next
          end

          row_offset = to_row - from_row
          col_offset = to_col - from_col

          # Row first
          begin
            move = []

            if row_offset != 0
              if row_offset < 0
                move << ['^', row_offset.abs]
              else
                move << ['v', row_offset.abs]
              end
            end

            if col_offset != 0
              if col_offset < 0
                move << ['<', col_offset.abs]
              else
                move << ['>', col_offset.abs]
              end
            end

            moves << move
          end

          # Col first
          begin
            move = []

            if col_offset != 0
              if col_offset < 0
                move << ['<', col_offset.abs]
              else
                move << ['>', col_offset.abs]
              end
            end

            if row_offset != 0
              if row_offset < 0
                move << ['^', row_offset.abs]
              else
                move << ['v', row_offset.abs]
              end
            end

            moves << move
          end

          # Walk the path and make sure we never hit an X
          moves.uniq.each do |move|
            row_pos = from_row
            col_pos = from_col

            ignore_move = false

            move.each do |(direction, movement)|
              movement.times do
                if direction == '^'
                  row_pos -= 1
                elsif direction == 'v'
                  row_pos += 1
                elsif direction == '<'
                  col_pos -= 1
                elsif direction == '>'
                  col_pos += 1
                end

                if panel.fetch(row_pos).fetch(col_pos) == 'X'
                  ignore_move = true
                end
              end
            end

            unless ignore_move
              result[from_cell] ||= {}
              result[from_cell][to_cell] ||= []
              result[from_cell][to_cell] << move
            end
          end
        end
      end
    end
  end

  pp result
end

# precompute_best_moves(NUMBER_PANEL)
# precompute_best_moves(DIRECTION_PANEL)


NUMBER_PANEL_MOVES = {"7"=>
                      {"7"=>[[]],
                       "4"=>[[['v', 1]]],
                       "1"=>[[['v', 2]]],
                       "8"=>[[['>', 1]]],
                       "5"=>[[['v', 1], ['>', 1]], [['>', 1], ['v', 1]]],
                       "2"=>[[['v', 2], ['>', 1]], [['>', 1], ['v', 2]]],
                       "0"=>[[['>', 1], ['v', 3]]],
                       "9"=>[[['>', 2]]],
                       "6"=>[[['v', 1], ['>', 2]], [['>', 2], ['v', 1]]],
                       "3"=>[[['v', 2], ['>', 2]], [['>', 2], ['v', 2]]],
                       "A"=>[[['>', 2], ['v', 3]]]},
                      "4"=>
                      {"7"=>[[['^', 1]]],
                       "4"=>[[]],
                       "1"=>[[['v', 1]]],
                       "8"=>[[['^', 1], ['>', 1]], [['>', 1], ['^', 1]]],
                       "5"=>[[['>', 1]]],
                       "2"=>[[['v', 1], ['>', 1]], [['>', 1], ['v', 1]]],
                       "0"=>[[['>', 1], ['v', 2]]],
                       "9"=>[[['^', 1], ['>', 2]], [['>', 2], ['^', 1]]],
                       "6"=>[[['>', 2]]],
                       "3"=>[[['v', 1], ['>', 2]], [['>', 2], ['v', 1]]],
                       "A"=>[[['>', 2], ['v', 2]]]},
                      "1"=>
                      {"7"=>[[['^', 2]]],
                       "4"=>[[['^', 1]]],
                       "1"=>[[]],
                       "8"=>[[['^', 2], ['>', 1]], [['>', 1], ['^', 2]]],
                       "5"=>[[['^', 1], ['>', 1]], [['>', 1], ['^', 1]]],
                       "2"=>[[['>', 1]]],
                       "0"=>[[['>', 1], ['v', 1]]],
                       "9"=>[[['^', 2], ['>', 2]], [['>', 2], ['^', 2]]],
                       "6"=>[[['^', 1], ['>', 2]], [['>', 2], ['^', 1]]],
                       "3"=>[[['>', 2]]],
                       "A"=>[[['>', 2], ['v', 1]]]},
                      "8"=>
                      {"7"=>[[['<', 1]]],
                       "4"=>[[['v', 1], ['<', 1]], [['<', 1], ['v', 1]]],
                       "1"=>[[['v', 2], ['<', 1]], [['<', 1], ['v', 2]]],
                       "8"=>[[]],
                       "5"=>[[['v', 1]]],
                       "2"=>[[['v', 2]]],
                       "0"=>[[['v', 3]]],
                       "9"=>[[['>', 1]]],
                       "6"=>[[['v', 1], ['>', 1]], [['>', 1], ['v', 1]]],
                       "3"=>[[['v', 2], ['>', 1]], [['>', 1], ['v', 2]]],
                       "A"=>[[['v', 3], ['>', 1]], [['>', 1], ['v', 3]]]},
                      "5"=>
                      {"7"=>[[['^', 1], ['<', 1]], [['<', 1], ['^', 1]]],
                       "4"=>[[['<', 1]]],
                       "1"=>[[['v', 1], ['<', 1]], [['<', 1], ['v', 1]]],
                       "8"=>[[['^', 1]]],
                       "5"=>[[]],
                       "2"=>[[['v', 1]]],
                       "0"=>[[['v', 2]]],
                       "9"=>[[['^', 1], ['>', 1]], [['>', 1], ['^', 1]]],
                       "6"=>[[['>', 1]]],
                       "3"=>[[['v', 1], ['>', 1]], [['>', 1], ['v', 1]]],
                       "A"=>[[['v', 2], ['>', 1]], [['>', 1], ['v', 2]]]},
                      "2"=>
                      {"7"=>[[['^', 2], ['<', 1]], [['<', 1], ['^', 2]]],
                       "4"=>[[['^', 1], ['<', 1]], [['<', 1], ['^', 1]]],
                       "1"=>[[['<', 1]]],
                       "8"=>[[['^', 2]]],
                       "5"=>[[['^', 1]]],
                       "2"=>[[]],
                       "0"=>[[['v', 1]]],
                       "9"=>[[['^', 2], ['>', 1]], [['>', 1], ['^', 2]]],
                       "6"=>[[['^', 1], ['>', 1]], [['>', 1], ['^', 1]]],
                       "3"=>[[['>', 1]]],
                       "A"=>[[['v', 1], ['>', 1]], [['>', 1], ['v', 1]]]},
                      "0"=>
                      {"7"=>[[['^', 3], ['<', 1]]],
                       "4"=>[[['^', 2], ['<', 1]]],
                       "1"=>[[['^', 1], ['<', 1]]],
                       "8"=>[[['^', 3]]],
                       "5"=>[[['^', 2]]],
                       "2"=>[[['^', 1]]],
                       "0"=>[[]],
                       "9"=>[[['^', 3], ['>', 1]], [['>', 1], ['^', 3]]],
                       "6"=>[[['^', 2], ['>', 1]], [['>', 1], ['^', 2]]],
                       "3"=>[[['^', 1], ['>', 1]], [['>', 1], ['^', 1]]],
                       "A"=>[[['>', 1]]]},
                      "9"=>
                      {"7"=>[[['<', 2]]],
                       "4"=>[[['v', 1], ['<', 2]], [['<', 2], ['v', 1]]],
                       "1"=>[[['v', 2], ['<', 2]], [['<', 2], ['v', 2]]],
                       "8"=>[[['<', 1]]],
                       "5"=>[[['v', 1], ['<', 1]], [['<', 1], ['v', 1]]],
                       "2"=>[[['v', 2], ['<', 1]], [['<', 1], ['v', 2]]],
                       "0"=>[[['v', 3], ['<', 1]], [['<', 1], ['v', 3]]],
                       "9"=>[[]],
                       "6"=>[[['v', 1]]],
                       "3"=>[[['v', 2]]],
                       "A"=>[[['v', 3]]]},
                      "6"=>
                      {"7"=>[[['^', 1], ['<', 2]], [['<', 2], ['^', 1]]],
                       "4"=>[[['<', 2]]],
                       "1"=>[[['v', 1], ['<', 2]], [['<', 2], ['v', 1]]],
                       "8"=>[[['^', 1], ['<', 1]], [['<', 1], ['^', 1]]],
                       "5"=>[[['<', 1]]],
                       "2"=>[[['v', 1], ['<', 1]], [['<', 1], ['v', 1]]],
                       "0"=>[[['v', 2], ['<', 1]], [['<', 1], ['v', 2]]],
                       "9"=>[[['^', 1]]],
                       "6"=>[[]],
                       "3"=>[[['v', 1]]],
                       "A"=>[[['v', 2]]]},
                      "3"=>
                      {"7"=>[[['^', 2], ['<', 2]], [['<', 2], ['^', 2]]],
                       "4"=>[[['^', 1], ['<', 2]], [['<', 2], ['^', 1]]],
                       "1"=>[[['<', 2]]],
                       "8"=>[[['^', 2], ['<', 1]], [['<', 1], ['^', 2]]],
                       "5"=>[[['^', 1], ['<', 1]], [['<', 1], ['^', 1]]],
                       "2"=>[[['<', 1]]],
                       "0"=>[[['v', 1], ['<', 1]], [['<', 1], ['v', 1]]],
                       "9"=>[[['^', 2]]],
                       "6"=>[[['^', 1]]],
                       "3"=>[[]],
                       "A"=>[[['v', 1]]]},
                      "A"=>
                      {"7"=>[[['^', 3], ['<', 2]]],
                       "4"=>[[['^', 2], ['<', 2]]],
                       "1"=>[[['^', 1], ['<', 2]]],
                       "8"=>[[['^', 3], ['<', 1]], [['<', 1], ['^', 3]]],
                       "5"=>[[['^', 2], ['<', 1]], [['<', 1], ['^', 2]]],
                       "2"=>[[['^', 1], ['<', 1]], [['<', 1], ['^', 1]]],
                       "0"=>[[['<', 1]]],
                       "9"=>[[['^', 3]]],
                       "6"=>[[['^', 2]]],
                       "3"=>[[['^', 1]]],
                       "A"=>[[]]}}


DIRECTION_PANEL_MOVES = {"<"=>
                         {"<"=>[[]],
                          "^"=>[[['>', 1], ['^', 1]]],
                          "v"=>[[['>', 1]]],
                          "A"=>[[['>', 2], ['^', 1]]],
                          ">"=>[[['>', 2]]]},
                         "^"=>
                         {"<"=>[[['v', 1], ['<', 1]]],
                          "^"=>[[]],
                          "v"=>[[['v', 1]]],
                          "A"=>[[['>', 1]]],
                          ">"=>[[['v', 1], ['>', 1]], [['>', 1], ['v', 1]]]},
                         "v"=>
                         {"<"=>[[['<', 1]]],
                          "^"=>[[['^', 1]]],
                          "v"=>[[]],
                          "A"=>[[['^', 1], ['>', 1]], [['>', 1], ['^', 1]]],
                          ">"=>[[['>', 1]]]},
                         "A"=>
                         {"<"=>[[['v', 1], ['<', 2]]],
                          "^"=>[[['<', 1]]],
                          "v"=>[[['v', 1], ['<', 1]], [['<', 1], ['v', 1]]],
                          "A"=>[[]],
                          ">"=>[[['v', 1]]]},
                         ">"=>
                         {"<"=>[[['<', 2]]],
                          "^"=>[[['^', 1], ['<', 1]], [['<', 1], ['^', 1]]],
                          "v"=>[[['<', 1]]],
                          "A"=>[[['^', 1]]],
                          ">"=>[[]]}}


Position = Struct.new(:row, :col)

def lowest_cost(code, directional_panel_count)
  current_cell = 'A'

  total_cost = 0

  code.each_char do |digit|
    best_cost = 2**63
    best_move = nil

    NUMBER_PANEL_MOVES.fetch(current_cell).fetch(digit).each do |move|
      cost = lowest_directional_cost(move + [['A', 1]], directional_panel_count)

      if cost < best_cost
        best_cost = cost
        best_move = move
      end
    end

    current_cell = digit

    total_cost += best_cost
  end

  total_cost
end


MEMO = {}

def lowest_directional_cost(desired_move, depth)
  if MEMO[[desired_move, depth]]
    return MEMO[[desired_move, depth]]
  end

  result = if depth == 0
             desired_move.map {|movement| movement[1]}.sum
           else
             current_cell = 'A'

             total_cost = 0

             desired_move.each do |movement|
               best_cost = 2**63
               best_move = nil

               DIRECTION_PANEL_MOVES.fetch(current_cell).fetch(movement[0]).each do |move|
                 cost = lowest_directional_cost(move + [['A', 1]], depth - 1) + (movement[1] - 1)

                 if cost < best_cost
                   best_cost = cost
                   best_move = move
                 end
               end

               total_cost += best_cost

               current_cell = movement[0]
             end

             total_cost
           end

  MEMO[[desired_move, depth]] = result

  result
end

def part1
  codes_to_enter = File.read("input_files/day21.txt").scan(/[0-9A]+/).to_a

  total_complexity = 0

  codes_to_enter.each do |code|
    cost = lowest_cost(code, 2)

    numeric = Integer(code.gsub(/[^0-9]/, ''))

    total_complexity += (cost * numeric)
  end

  puts "Part 1: Total complexity: #{total_complexity}"
end


def part2
  codes_to_enter = File.read("input_files/day21.txt").scan(/[0-9A]+/).to_a

  total_complexity = 0

  codes_to_enter.each do |code|
    cost = lowest_cost(code, 25)

    numeric = Integer(code.gsub(/[^0-9]/, ''))

    total_complexity += (cost * numeric)
  end

  puts "Part 2: Total complexity: #{total_complexity}"
end

part1
part2
